File: frontend/lib/types/product.ts
/**
 * Product domain types.
 */

/** Per-product breakdown of how demand_score/opportunity_score were computed. */
export interface ScoreBreakdownWeights {
  rating: number;
  demand: number;
  trend: number;
}

export interface ScoreBreakdown {
  /** rating (0-5) normalized to a 0-100 scale. */
  rating_norm: number;
  /** Same value as Product.demand_score, echoed here for the formula substitution. */
  demand_score: number;
  /** The product's category trend score (0-100) at generation time. */
  trend_score: number;
  weights: ScoreBreakdownWeights;
  /** Human-readable demand formula, e.g. "log10(review_count+1)/log10(5001)*100" */
  demand_formula: string;
  /** Human-readable opportunity formula. */
  opportunity_formula: string;
}

export interface Product {
  id: number;
  name: string;
  category_id: number;
  price: number;
  rating: number;
  review_count: number;
  features: string[];
  /** Canonical material tags (e.g. ["Wood", "Metal"]); may be empty. */
  material: string[];
  /** Amazon ASIN, or null in mock-data mode. */
  asin: string | null;
  /** Amazon product URL, or null in mock-data mode. */
  url: string | null;
  /** Composite score (0-100) indicating whitespace / opportunity for a competing product. */
  opportunity_score: number;
  /** Composite score (0-100) indicating current market demand. */
  demand_score: number;
  score_breakdown: ScoreBreakdown;
}

/** Columns that the Product Ranking table can be sorted by. */
export type ProductSortField =
  | "name"
  | "price"
  | "rating"
  | "review_count"
  | "opportunity_score"
  | "demand_score";

export type SortOrder = "asc" | "desc";

/** One material facet bucket: the material value plus how many products have it. */
export interface ProductFacetMaterial {
  value: string;
  count: number;
}

/** Response shape for GET /api/products/facets */
export interface ProductFacets {
  price_min: number;
  price_max: number;
  rating_min: number;
  rating_max: number;
  materials: ProductFacetMaterial[];
}

File: backend/app/schemas/product.py
"""Pydantic schemas for the Product resource."""

from pydantic import BaseModel, Field

from app.schemas.common import ORMBase


class ScoreBreakdown(BaseModel):
    """
    The inputs and formula behind a product's derived scores, returned
    alongside every product so the UI can show *why* a product scored
    what it did rather than presenting an unexplained number.

    Product managers act on these rankings, so an opaque score is not
    good enough - the breakdown makes the model auditable at a glance.
    Weights and formulas are echoed from product_service (the single
    source of truth) instead of being restated in the frontend, so the
    displayed explanation can never drift from the actual computation.
    """

    rating_norm: float = Field(..., description="rating / 5 * 100, i.e. satisfaction on a 0-100 scale")
    demand_score: float = Field(..., description="Review-volume-derived demand, 0-100")
    trend_score: float = Field(..., description="The category's current market trend score, 0-100")
    weights: dict[str, float] = Field(..., description="Weight applied to each component")
    demand_formula: str = Field(..., description="How demand_score is derived from review_count")
    opportunity_formula: str = Field(..., description="How opportunity_score combines the components")


class ProductRead(ORMBase):
    """Response schema for a Product row, including derived scoring fields."""

    id: int
    name: str
    category_id: int
    price: float
    rating: float
    review_count: int
    features: list[str]
    # Canonical English material tags, e.g. ["Solid Wood", "Steel"]. May be
    # empty when the listing text names no recognizable material.
    material: list[str] = Field(default_factory=list)
    # Null for provider-synthesized (mock) products, which have no real
    # marketplace listing to link to.
    asin: str | None = None
    url: str | None = None
    opportunity_score: float
    demand_score: float
    # Computed per-request (not stored) - see product_service.build_score_breakdown.
    score_breakdown: ScoreBreakdown | None = None


class ProductFacets(BaseModel):
    """
    Filter metadata for GET /api/products/facets.

    Lets the UI build price/rating/material controls from what is actually
    present in the data (optionally scoped to one category) instead of
    hardcoding ranges and material lists that would go stale the moment
    the catalogue is re-seeded from a live provider.
    """

    price_min: float
    price_max: float
    rating_min: float
    rating_max: float
    materials: list[dict] = Field(
        default_factory=list,
        description='Available materials with counts, e.g. [{"value": "Wood", "count": 12}]',
    )
